vsintegration/src/FSharp.Editor/LanguageService/LanguageService.fs

```diff
@@ -118,7 +119,8 @@ type internal FSharpProjectOptionsManager
     (
         checkerProvider: FSharpCheckerProvider,
         [<Import(typeof<VisualStudioWorkspace>)>] workspace: VisualStudioWorkspaceImpl,
-        [<Import(typeof<SVsServiceProvider>)>] serviceProvider: System.IServiceProvider
+        [<Import(typeof<SVsServiceProvider>)>] serviceProvider: System.IServiceProvider,
+        [<Import(typeof<ISettings>)>] _settings: ISettings // This is necessary so that the Settings global doesn't explode.
```

Comment:
It's not clear how this is related to the rest of the change.

Reply:
It's a bit complicated. We have a `Settings` type that relies on the implementation of `ISettings`s' constructor to be called. We import it here to force the constructor to be called at this point. Unfortunately we do a similar thing already in the `FSharpPackage`'s `Initialize` method, but since `Initialize` hasn't been called yet until a document has been opened, we have to do this in order for it to work. 

This is absolutely not ideal and a horrific pattern. We need to re-design how all of this flows through our language service.

Reply:
If `FSharpPackage` hasn't been initialized, I expect lots of horrible things to happen.  I would look into fixing that first.

Reply:
@pilchie, the updateprojectinfo relies on access to the F# VS Settings.  This causes them to be initialized when the CommandLineOptionsEvent arrived.  The CommandLine options event may fire before the package is fully initialized.

Reply:
> The CommandLine options event may fire before the package is fully initialized.

We should fix that.

Reply:
FYI, we use `Settings` type when trying to update project information.

Reply:
To be clear. `FSharpPackage` gets instantiated through MEF when we load a solution. Its just the `Initialize` method doesn't get called until we open a document. Talking with @jasonmalinowski , this is intended and correct behavior.

Reply:
Ideally the correct fix for all of this, would be to change how we hook up the project-system and Workspace to our IDE tooling. I believe the best way to do this is to implement F# compilation options for the Workspace. This way in our tooling, when we look at a document, we can ask for the Project, then ask for the project compilation options using the Workspace APIs. Today, we have to manage this ourselves based on these callbacks in our language service.

Reply:
That's unlikely to be feasible any time soon, since the commandline arguments types are strongly connected to the C#/VB types, and include compiler concepts that don't apply to F#.

Regardless, it wouldn't help this initialization piece.  In the short term though, I'm not sure why we couldn't call `IVsShell::LoadPackage` (or `QueryService` for a service profferred from the language service) from the project system to ensure the language service is initialized before we start using it.

It's historically been very hard to reason about VS packages that might get loaded by MEF but not have `Initialize` called on them.  If that's happening in F#, my guess is that this is far from the only bug that happens as a result, and it would be better to fix the underlying issue.

Can you look at how the C# Language Service package gets initialized in a similar codepath and explain why that doesn't happen for F#?

Reply:
For a long-term solution, I really think we should explore what could possibly be done for the commandline arguments. @jasonmalinowski thinks we might be able to do this. We should all talk about this together.

The `ISettings` initialization is perfectly fine to happen here; we just have a bad pattern where a global structure relies on its initialization. In a separate PR, we can remove the global structure that and just use the `ISettings` instance when we need it.

Reply:
Regardless if we use `LoadPackage` or not, we should take this PR as using `CurrentSolution` is not correct here anyways.

Reply:
After checking with @jasonmalinowski , calling `IVsShell::LoadPackage` seems reasonable to do.

Reply:
I think I found where to do it: https://github.com/dotnet/roslyn/blob/0959f9d826116a47a6da267cd1420fc803568b28/src/VisualStudio/Core/Impl/ProjectSystem/CPS/CPSProjectFactory.cs#L122
Seems F# is not in that list. :)